third_party/ibis/ibis_mssql/compiler.py

```diff
@@ -489,18 +489,47 @@ class MSSQLExprTranslator(alch.AlchemyExprTranslator):
     _registry = _operation_registry
     _rewrites = alch.AlchemyExprTranslator._rewrites.copy()
     _type_map = alch.AlchemyExprTranslator._type_map.copy()
+    # The following MSSQL 2022 types are not supported: 'XML', 'SQL_VARIANT', 'SYSNAME', 'HIERARCHYID',
     _type_map.update(
         {
+            # Exact numerics
             dt.Boolean: mssql.BIT,
             dt.Int8: mssql.TINYINT,
+            dt.Int16: mssql.SMALLINT,
             dt.Int32: mssql.INTEGER,
             dt.Int64: mssql.BIGINT,
-            dt.Float: mssql.REAL,
-            dt.Double: mssql.REAL,
-            dt.String: mssql.VARCHAR,
+            dt.Float16: mssql.FLOAT,
+            dt.Float32: mssql.FLOAT,
+            dt.Float64: mssql.REAL,
```

Comment:
Have we got Float and REAL the wrong way around here?

According to SQL Server docs: https://learn.microsoft.com/en-us/sql/t-sql/data-types/float-and-real-transact-sql?view=sql-server-ver16

"real is float(24)"

Which, I think, means real holds 32 bit values and float can go up to 64 bit.

Reply:
Same as my other answer about the reference for my implementation (https://github.com/ibis-project/ibis/blob/master/ibis/backends/mssql/datatypes.py#L96) and actually this part really confused me, specially having a duplicate mapping for FLOAT.  

So you suggest this?

```suggestion
            dt.Float64: mssql.FLOAT,
            dt.Float32: mssql.REAL,
```

Reply:
In the code right above the code you linked to Ibis maps like this:
```
    # Approximate numerics
    'FLOAT': dt.Float64,
    'REAL': dt.Float32,
```

But then they do the opposite right below (as you pointed out). Again, it looks wrong to me but I might be missing something. Perhaps it is safest to do as you've done until we run into an issue. We could easily test this using row validation on a big 64 bit value in a FLOAT column (I'm not suggesting you do this now but we could come back to it - or raise an issue to remind us to test it).

Perhaps the best way forward is to add a comment to our code saying we're not convinced it is right but have gone with it based on Ibis code?

Reply:
I wonder if this could be related: https://github.com/GoogleCloudPlatform/professional-services-data-validator/issues/719